enhancements/machine-config/enable-evented-pleg-in-openshift.md

```diff
@@ -0,0 +1,196 @@
+	// +required
+	Spec NodeSpec `json:"spec"`
+
+	// status holds observed values.
+	// +optional
+	Status NodeStatus `json:"status"`
+}
+
+
+type NodeSpec struct {
+	// CgroupMode
+
+	// WorkerLatencyProfile
+
+	// EnablePodEvents determines whether to enable the evented pleg feature to
+	// send the container events from the CRI-O through gRPC to the Kubelet
+	// +optional
+	EnablePodEvents bool `json:"enablePodEvents,omitempty"`
+}
+
+```
+The following machine configs are newly rendered or updated(if already present) when there is an update made to the above mentioned `EnablePodEvents` field.
+- `97-master-generated-kubelet`
+- `97-worker-generated-kubelet`
+- `97-master-generated-crio-enable-pod-events`
+- `97-worker-generated-crio-enable-pod-events`
+The kubelet and the cri-o configurations on the nodes are updated as follows when `EnablePodEvents` is set to `true`:
+1. Kubelet Configuration:
+```json
+{
+  "kind": "KubeletConfiguration",
+  "apiVersion": "kubelet.config.k8s.io/v1beta1",
+  "featureGates": {
+    "EventedPLEG": true,
+  },
+}
+```
+2. CRI-O Configuration:
+```yaml
+[crio]
+  [crio.runtime]
+    enable_pod_events = true
+```
+
+### Risks and Mitigations
+
+- PLEG is very core to the container status handling in the kubelet. Hence any miscalculation there would result in unpredictable behaviour not just for the node but for an entire cluster.
+  - To reduce the risk of regression, this feature initially will be available only as an opt-in.
+  - Users can disable this feature to make kubelet use existing relisting based PLEG.
+- Another risk is the CRI implementation could have a buggy event emitting system, and miss pod lifecycle events.
+  - A mitigation is a `kube_pod_missed_events` metric, which the Kubelet could report when a lifecycle event is registered that wasn't triggered by an event, but rather by changes of state between lists.
+  - While using the Evented implementation, the periodic relisting functionality would still be used with an increased interval which should work as a fallback mechanism for missed events in case of any disruptions.
+  - Evented PLEG will need to update global cache timestamp periodically in order to make sure pod workers don't get stuck at [GetNewerThan](https://github.com/kubernetes/kubernetes/blob/4a894be926adfe51fd8654dcceef4ece89a4259f/pkg/kubelet/pod_workers.go#L924) in case Evented PLEG misses the event for any unforeseen reason.
+
+### Drawbacks
+
+This feature requires changes in the code base of the PLEG module that adjusts the container runtime state, accordngly updates the pod cache.
+Kubelet depends on the events generated by the container runtimes(Evented PLEG) to update the pod cache.
+Sometimes, there is a chance of events getting missed and hence relying alone on the evented pleg, without having sufficient benchmark data, is not recommended.
+
+One more minor drawback of this feature is that the node gets updated/rebooted twice with the two newly rendered machine configs
```

Comment:
I think we should make it "node may gets updated/rebooted twice with the two newly rendered machine configs" because this can happen only in certain scenario as mentioned by Jerry in [comment](https://github.com/openshift/enhancements/pull/1368/files#r1163447085).

Reply:
Addressed, Thanks!